File: leetcode/go/1717.maximum-score-from-removing-substrings.go
func maximumGain(s string, x int, y int) int {
    if x > y {
        return gain([]rune(s), []rune("ab"), []rune("ba"), x, y)
    } else {
        return gain([]rune(s), []rune("ba"), []rune("ab"), y, x)
    }
}

func gain(s []rune, sub1 []rune, sub2 []rune, p1 int, p2 int) int {
    st := make(stack, 0, len(s))
    g := 0
    for _, c := range s {
        if len(st) > 0 && c == sub1[1] && st.Peek() == sub1[0] {
            st, _ = st.Pop()
            g += p1
        } else {
            st = st.Push(c)
        }
    }
    st2 := make(stack, 0, len(st))
    for _, c := range st {
        if len(st2) > 0 && c == sub2[1] && st2.Peek() == sub2[0] {
            st2, _ = st2.Pop()
            g += p2
        } else {
            st2 = st2.Push(c)
        }
    }
    return g
}

type stack []rune

func (s stack) Push(v rune) stack {
    return append(s, v)
}

func (s stack) Pop() (stack, rune) {
    l := len(s)
    return  s[:l-1], s[l-1]
}

func (s stack) Peek() (rune) {
    return s[len(s)-1]
}

File: leetcode/rust-leetcode/src/1717.maximum-score-from-removing-substrings.rs
impl Solution {
    pub fn maximum_gain(s: String, x: i32, y: i32) -> i32 {
        let mut sb = s.into_bytes();
        
        let sub1 = [b'a', b'b'];
        let sub2 = [b'b', b'a'];

        let gain = |s1: [u8; 2], s2: [u8; 2], p1: i32, p2: i32| -> i32 {
            let mut stack1 = Vec::new();
            let mut p = 0;
            for &c in sb.iter() {
                if c == s1[1] && !stack1.is_empty() && *stack1.last().unwrap() == s1[0] {
                    p += p1;
                    stack1.pop();
                } else {
                    stack1.push(c);
                }
            }
            let mut stack2 = Vec::new();
            for &c in stack1.iter() {
                if c == s2[1] && !stack2.is_empty() && *stack2.last().unwrap() == s2[0] {
                    p += p2;
                    stack2.pop();
                } else {
                    stack2.push(c);
                }
            }
            p
        };
        
        if x > y {
            gain(sub1, sub2, x, y)
        } else {
            gain(sub2, sub1, y, x)
        }
    }
}


